File: backend/app/schemas/money_trust.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel


MoneyTrustState = Literal[
    "confirmed",
    "provisional",
    "estimated",
    "opportunity",
    "test_only",
    "blocked",
]

MoneyImpactKind = Literal[
    "confirmed_loss",
    "probable_loss",
    "probable_risk",
    "blocked_cash",
    "lost_sales_risk",
    "opportunity",
    "estimated_opportunity",
    "blocked_revenue",
    "data_blocker",
    "data_blocked",
    "test_only",
    "informational",
]


class MoneyTrustInfo(BaseModel):
    state: MoneyTrustState
    impact_kind: MoneyImpactKind
    display_label: str
    amount_label: str
    show_as_confirmed_money: bool = False
    seller_visible_by_default: bool = True
    reason: str = ""
    evidence_trust_state: MoneyTrustState | None = None
    impact_trust_state: MoneyTrustState | None = None
    saved_money_claimed: bool = False


def _norm(value: Any) -> str:
    return str(value or "").strip().lower()


def _has_any(text: str, needles: set[str]) -> bool:
    return any(needle in text for needle in needles)


def classify_money_trust(
    *,
    value: Any = None,
    value_type: str | None = None,
    confidence: str | None = None,
    impact_type: str | None = None,
    trust_state: str | None = None,
    financial_final: bool | None = None,
    source_module: str | None = None,
    source_table: str | None = None,
    source_endpoint: str | None = None,
    action_type: str | None = None,
    payload: dict[str, Any] | None = None,
    affected_amount: float | int | None = None,
    affected_revenue: float | int | None = None,
) -> MoneyTrustInfo:
    """Classify visible money so UI never turns weak signals into real losses."""

    payload = payload or {}
    text = " ".join(
        _norm(part)
        for part in (
            confidence,
            impact_type,
            trust_state,
            source_module,
            source_table,
            source_endpoint,
            action_type,
            payload.get("trust_state"),
            payload.get("runtime_mode"),
            payload.get("source"),
            payload.get("code"),
        )
    )
    module = _norm(source_module)
    impact = _norm(impact_type)
    conf = _norm(confidence)
    trust = _norm(trust_state or payload.get("trust_state"))
    evidence_trust = _trust_state_for_contract(
        payload.get("evidence_trust_state") or conf or trust
    )
    impact_trust = _trust_state_for_contract(
        payload.get("impact_trust_state") or trust or conf
    )
    action = _norm(action_type)
    if action in {"low_stock_risk", "fast_stock_depletion"}:
        impact = "lost_sales_risk"
    elif action in {"overstock_slow_moving", "dead_stock"} and impact in {
        "",
        "confirmed_loss",
        "opportunity",
    }:
        impact = "blocked_cash"
    numeric_value = _to_float(value)
    amount = _to_float(affected_amount)
    revenue = _to_float(affected_revenue)

    if (
        trust in {"test_only", "test", "mock", "demo"}
        or conf in {"test_only", "test"}
        or payload.get("test_only") is True
        or payload.get("beta") is True
        or module == "grouping_beta"
    ):
        return MoneyTrustInfo(
            state="test_only",
            impact_kind="test_only",
            display_label="Тестовый сигнал",
            amount_label="Не деньги продавца",
            show_as_confirmed_money=False,
            seller_visible_by_default=False,
            reason="Generated/test-only signal is hidden from the default seller view.",
            evidence_trust_state=evidence_trust,
            impact_trust_state="test_only",
            saved_money_claimed=False,
        )

    if (
        conf == "blocked"
        or impact == "data_blocker"
        or trust in {"blocked", "data_blocked"}
    ):
        blocked_revenue_only = (amount in {None, 0.0}) and revenue not in {None, 0.0}
        return MoneyTrustInfo(
            state="blocked",
            impact_kind="blocked_revenue" if blocked_revenue_only else "data_blocker",
            display_label="Данные заблокированы",
            amount_label="Блокированная выручка"
            if blocked_revenue_only
            else "Не хватает данных",
            show_as_confirmed_money=False,
            seller_visible_by_default=True,
            reason="Required source data is missing or invalid, so final money cannot be computed.",
            evidence_trust_state="blocked",
            impact_trust_state="blocked",
            saved_money_claimed=False,
        )

    if impact == "lost_sales_risk":
        risk_trust: MoneyTrustState = (
            impact_trust
            if impact_trust in {"provisional", "estimated"}
            else "provisional"
        )
        return MoneyTrustInfo(
            state=risk_trust,
            impact_kind="lost_sales_risk",
            display_label="Риск потери продаж",
            amount_label="Риск потери продаж",
            show_as_confirmed_money=False,
            seller_visible_by_default=True,
            reason="Low stock can create a sales-loss risk, but it is not a confirmed loss until after-data proves missed sales.",
            evidence_trust_state=evidence_trust,
            impact_trust_state=risk_trust,
            saved_money_claimed=False,
        )

    if impact == "blocked_cash":
        blocked_cash_trust: MoneyTrustState = (
            "confirmed"
            if financial_final is True and impact_trust == "confirmed"
            else "estimated"
        )
        return MoneyTrustInfo(
            state=blocked_cash_trust,
            impact_kind="blocked_cash",
            display_label="Замороженные деньги",
            amount_label="Замороженные деньги",
            show_as_confirmed_money=False,
            seller_visible_by_default=True,
            reason="Overstock ties up cash in inventory; it is not a confirmed realized loss.",
            evidence_trust_state=evidence_trust,
            impact_trust_state=blocked_cash_trust,
            saved_money_claimed=False,
        )

    if impact in {"probable_loss", "probable_risk"}:
        risk_trust: MoneyTrustState = (
            impact_trust
            if impact_trust in {"provisional", "estimated"}
            else "estimated"
        )
        return MoneyTrustInfo(
            state=risk_trust,
            impact_kind="probable_loss"
            if impact == "probable_loss"
            else "probable_risk",
            display_label="Вероятный убыток"
            if impact == "probable_loss"
            else "Вероятный риск",
            amount_label="Вероятный убыток"
            if impact == "probable_loss"
            else "Вероятный риск",
            show_as_confirmed_money=False,
            seller_visible_by_default=True,
            reason="The impact is a risk estimate and must not be shown as a confirmed loss.",
            evidence_trust_state=evidence_trust,
            impact_trust_state=risk_trust,
            saved_money_claimed=False,
        )

    finance_source = _has_any(
        text, {"finance", "financial", "realization", "report", "wb_realization"}
    )
    if impact == "confirmed_loss" and (
        financial_final is True or conf == "confirmed" or finance_source
    ):
        return MoneyTrustInfo(
            state="confirmed",
            impact_kind="confirmed_loss",
            display_label="Подтверждённый убыток",
            amount_label="Подтверждённый убыток",
            show_as_confirmed_money=True,
            seller_visible_by_default=True,
            reason="Confirmed by WB finance/report rows or user-confirmed source data.",
            evidence_trust_state=evidence_trust,
            impact_trust_state="confirmed",
            saved_money_claimed=False,
        )

    if module == "checker" or _has_any(
        text, {"checker", "card_quality", "content", "photo", "description"}
    ):
        return MoneyTrustInfo(
            state="opportunity",
            impact_kind="estimated_opportunity",
            display_label="Оценочная возможность",
            amount_label="Оценочная возможность",
            show_as_confirmed_money=False,
            seller_visible_by_default=True,
            reason="Checker/content improvements are conversion opportunities, not confirmed financial losses.",
            evidence_trust_state=evidence_trust,
            impact_trust_state="opportunity",
            saved_money_claimed=False,
        )

    if impact in {"opportunity", "estimated_opportunity"}:
        return MoneyTrustInfo(
            state="opportunity"
            if conf not in {"estimated", "estimate"}
            else "estimated",
            impact_kind="opportunity"
            if impact == "opportunity"
            else "estimated_opportunity",
            display_label="Возможность роста"
            if impact == "opportunity"
            else "Оценочная возможность",
            amount_label="Возможность роста"
            if impact == "opportunity"
            else "Оценочная возможность",
            show_as_confirmed_money=False,
            seller_visible_by_default=True,
            reason="The value is an expected effect, not confirmed money already lost.",
            evidence_trust_state=evidence_trust,
            impact_trust_state="opportunity"
            if conf not in {"estimated", "estimate"}
            else "estimated",
            saved_money_claimed=False,
        )

    if conf in {"estimated", "estimate"} or _has_any(
        text, {"estimated", "estimate", "allocation", "model"}
    ):
        return MoneyTrustInfo(
            state="estimated",
            impact_kind="probable_risk"
            if (numeric_value or 0) < 0
            else "estimated_opportunity",
            display_label="Оценка",
            amount_label="Оценочная возможность"
            if (numeric_value or 0) >= 0
            else "Вероятный риск",
            show_as_confirmed_money=False,
            seller_visible_by_default=True,
            reason="The value is inferred, allocated, or model-based.",
            evidence_trust_state=evidence_trust,
            impact_trust_state="estimated",
            saved_money_claimed=False,
        )

    if financial_final is True or (conf == "confirmed" and finance_source):
        return MoneyTrustInfo(
            state="confirmed",
            impact_kind="informational",
            display_label="Подтверждённые деньги",
            amount_label="Подтверждённые деньги",
            show_as_confirmed_money=True,
            seller_visible_by_default=True,
            reason="Confirmed by WB finance/report rows or user-confirmed source data.",
            evidence_trust_state=evidence_trust,
            impact_trust_state="confirmed",
            saved_money_claimed=False,
        )

    return MoneyTrustInfo(
        state="provisional",
        impact_kind="probable_risk"
        if impact in {"probable_loss", "probable_risk"}
        else "informational",
        display_label="Предварительно",
        amount_label="Вероятный риск"
        if impact in {"probable_loss", "probable_risk"}
        else "Предварительные деньги",
        show_as_confirmed_money=False,
        seller_visible_by_default=True,
        reason="Operational money before final finance/source confirmation.",
        evidence_trust_state=evidence_trust,
        impact_trust_state="provisional",
        saved_money_claimed=False,
    )


def _trust_state_for_contract(value: Any) -> MoneyTrustState:
    normalized = _norm(value)
    if normalized in {"confirmed", "trusted", "final", "high"}:
        return "confirmed"
    if normalized in {"estimated", "estimate", "medium"}:
        return "estimated"
    if normalized in {"opportunity", "chance"}:
        return "opportunity"
    if normalized in {"test_only", "test", "mock", "demo"}:
        return "test_only"
    if normalized in {"blocked", "data_blocked", "data_blocker"}:
        return "blocked"
    return "provisional"


def _to_float(value: Any) -> float | None:
    if value is None or value == "":
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None

File: frontend/src/lib/money-trust.ts
export type MoneyTrustState =
  | "confirmed"
  | "provisional"
  | "estimated"
  | "opportunity"
  | "test_only"
  | "blocked"
  | string;

export type MoneyImpactKind =
  | "confirmed_loss"
  | "probable_loss"
  | "probable_risk"
  | "blocked_cash"
  | "lost_sales_risk"
  | "opportunity"
  | "estimated_opportunity"
  | "blocked_revenue"
  | "data_blocker"
  | "data_blocked"
  | "system_warning"
  | "test_only"
  | "informational"
  | string;

export interface MoneyTrustInfo {
  state: MoneyTrustState;
  impact_kind: MoneyImpactKind;
  display_label: string;
  amount_label: string;
  show_as_confirmed_money?: boolean;
  seller_visible_by_default?: boolean;
  reason?: string;
  evidence_trust_state?: MoneyTrustState | null;
  impact_trust_state?: MoneyTrustState | null;
  saved_money_claimed?: boolean;
  [key: string]: unknown;
}

const STATE_FALLBACK_LABEL: Record<string, { display_label: string; amount_label: string; impact_kind: string }> = {
  confirmed: { display_label: "Подтверждено", amount_label: "Подтверждённые деньги", impact_kind: "informational" },
  provisional: { display_label: "Предварительно", amount_label: "Вероятный риск", impact_kind: "probable_risk" },
  estimated: { display_label: "Оценка", amount_label: "Оценочная возможность", impact_kind: "estimated_opportunity" },
  opportunity: { display_label: "Возможность", amount_label: "Возможность роста", impact_kind: "opportunity" },
  test_only: { display_label: "Тестовое правило", amount_label: "Не деньги продавца", impact_kind: "test_only" },
  blocked: { display_label: "Не хватает данных", amount_label: "Не хватает данных", impact_kind: "data_blocker" },
};

const IMPACT_FALLBACK_LABEL: Record<string, { display_label: string; amount_label: string; impact_kind: string }> = {
  confirmed_loss: { display_label: "Подтверждённый убыток", amount_label: "Подтверждённый убыток", impact_kind: "confirmed_loss" },
  probable_loss: { display_label: "Вероятный убыток", amount_label: "Вероятный убыток", impact_kind: "probable_loss" },
  probable_risk: { display_label: "Вероятный риск", amount_label: "Вероятный риск", impact_kind: "probable_risk" },
  blocked_cash: { display_label: "Замороженные деньги", amount_label: "Замороженные деньги", impact_kind: "blocked_cash" },
  lost_sales_risk: { display_label: "Риск потери продаж", amount_label: "Риск потери продаж", impact_kind: "lost_sales_risk" },
  opportunity: { display_label: "Возможность роста", amount_label: "Возможность роста", impact_kind: "opportunity" },
  estimated_opportunity: { display_label: "Оценочная возможность", amount_label: "Оценочная возможность", impact_kind: "estimated_opportunity" },
  data_blocker: { display_label: "Не хватает данных", amount_label: "Не хватает данных", impact_kind: "data_blocker" },
  data_blocked: { display_label: "Данные заблокированы", amount_label: "Данные заблокированы", impact_kind: "data_blocked" },
  system_warning: { display_label: "Системное предупреждение", amount_label: "Системное предупреждение", impact_kind: "system_warning" },
  test_only: { display_label: "Тестовое правило", amount_label: "Не деньги продавца", impact_kind: "test_only" },
};

export function coerceMoneyTrust(value: unknown): MoneyTrustInfo | null {
  if (!value || typeof value !== "object" || Array.isArray(value)) return null;
  const obj = value as Record<string, unknown>;
  if (obj.money_trust) return coerceMoneyTrust(obj.money_trust);
  const state = typeof obj.state === "string" ? obj.state : null;
  if (!state) return null;
  const impactKind = typeof obj.impact_kind === "string" ? obj.impact_kind : null;
  const fallback =
    (impactKind ? IMPACT_FALLBACK_LABEL[impactKind] : null) ??
    STATE_FALLBACK_LABEL[state] ??
    STATE_FALLBACK_LABEL.provisional;
  return {
    state,
    impact_kind: impactKind ?? fallback.impact_kind,
    display_label: typeof obj.display_label === "string" ? obj.display_label : fallback.display_label,
    amount_label: typeof obj.amount_label === "string" ? obj.amount_label : fallback.amount_label,
    show_as_confirmed_money: obj.show_as_confirmed_money === true,
    seller_visible_by_default: obj.seller_visible_by_default !== false,
    reason: typeof obj.reason === "string" ? obj.reason : undefined,
    evidence_trust_state: typeof obj.evidence_trust_state === "string" ? obj.evidence_trust_state : undefined,
    impact_trust_state: typeof obj.impact_trust_state === "string" ? obj.impact_trust_state : state,
    saved_money_claimed: obj.saved_money_claimed === true,
    ...obj,
  };
}

export function moneyTrustFrom(...values: unknown[]): MoneyTrustInfo {
  for (const value of values) {
    const trust = coerceMoneyTrust(value);
    if (trust) return trust;
  }
  return {
    state: "provisional",
    impact_kind: "probable_risk",
    display_label: "Предварительно",
    amount_label: "Вероятный риск",
    show_as_confirmed_money: false,
    seller_visible_by_default: true,
    reason: "Классификатор доверия к деньгам не передан.",
    evidence_trust_state: "provisional",
    impact_trust_state: "provisional",
    saved_money_claimed: false,
  };
}

export function moneyTrustTone(trust: MoneyTrustInfo | null | undefined): string {
  const kind = String(trust?.impact_kind ?? trust?.state ?? "").toLowerCase();
  if (kind === "confirmed_loss") return "border-destructive/35 bg-destructive/10 text-destructive";
  if (kind === "probable_loss" || kind === "probable_risk" || kind === "lost_sales_risk") return "border-amber-500/35 bg-amber-500/10 text-amber-800 dark:text-amber-300";
  if (kind === "blocked_cash") return "border-orange-500/35 bg-orange-500/10 text-orange-800 dark:text-orange-300";
  if (kind === "test_only" || trust?.state === "test_only") return "border-slate-500/30 bg-slate-500/10 text-slate-700 dark:text-slate-300";
  if (kind === "blocked_revenue" || kind === "data_blocked" || kind === "data_blocker" || trust?.state === "blocked") return "border-amber-500/35 bg-amber-500/10 text-amber-800 dark:text-amber-300";
  if (kind === "estimated_opportunity" || kind === "opportunity" || trust?.state === "opportunity") return "border-sky-500/30 bg-sky-500/10 text-sky-700 dark:text-sky-300";
  if (trust?.state === "confirmed") return "border-emerald-500/30 bg-emerald-500/10 text-emerald-700 dark:text-emerald-300";
  return "border-amber-500/30 bg-amber-500/10 text-amber-800 dark:text-amber-300";
}

export function isSellerVisibleMoneyTrust(...values: unknown[]): boolean {
  return moneyTrustFrom(...values).seller_visible_by_default !== false;
}

export function evidenceTrustStateFrom(...values: unknown[]): MoneyTrustState | null {
  for (const value of values) {
    if (!value || typeof value !== "object" || Array.isArray(value)) continue;
    const obj = value as Record<string, unknown>;
    const nested = evidenceTrustStateFrom(obj.money_trust);
    if (nested) return nested;
    if (typeof obj.evidence_trust_state === "string" && obj.evidence_trust_state.trim()) {
      return obj.evidence_trust_state;
    }
    if (Array.isArray(obj.input_facts) && obj.input_facts.length > 0) {
      const factStates = obj.input_facts
        .map((fact) =>
          fact && typeof fact === "object" && !Array.isArray(fact)
            ? String((fact as Record<string, unknown>).trust_state ?? "").trim()
            : "",
        )
        .filter(Boolean);
      if (factStates.length > 0 && factStates.every((state) => state === "confirmed")) {
        return "confirmed";
      }
    }
    if (typeof obj.confidence === "string" && obj.confidence.trim()) {
      return obj.confidence;
    }
  }
  return null;
}

export function evidenceTrustLabel(value: unknown): string | null {
  const state = String(value ?? "").trim().toLowerCase();
  if (!state) return null;
  if (state === "confirmed" || state === "trusted" || state === "final") return "Данные подтверждены";
  if (state === "estimated" || state === "estimate") return "Данные оценочные";
  if (state === "provisional") return "Данные предварительные";
  if (state === "blocked" || state === "data_blocked" || state === "data_blocker") return "Данных не хватает";
  if (state === "test_only" || state === "test") return "Тестовые данные";
  return null;
}
